### PR TITLE
#5394: Add functional version of Mamba architecture

### PR DESCRIPTION
**Pull request by kpaigwar**

### CODE REVIEW

**Review by arakhmati — CHANGES_REQUESTED** [review]
If you want to put it under `models/demos`, then it has to be re-written using `ttnn`. Without any `import tt_lib`.

If you don't want to update the model right now to use `ttnn`, then please put it under `models/experimental`

**Review by eyonland — CHANGES_REQUESTED** [review]
Any new models going into models/demo needs to have run_performance.sh updated where the run_perf_models() is updated as well as the run_device_perf_models()

**Comment by kpaigwar** [comment]
Right now, there is no immediate plan to convert it using ttnn. I will move it to models/experimental. @eyonland this is our functional version, not optimized for performance. Does models in experimental also requires run_performance.sh?

File: models/demos/mamba/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Mamba Model Demo
+
+## Run on CPU
+```
+python demo/demo.py "What is the capital of France?" "What is the capital of Germany?" --model cpu
+```
+
+## Run on Wormhole
+```
+python demo/demo.py "What is the capital of France?" "What is the capital of Germany?" --model cpu
```
Comment (esmalTT) [inline models/demos/mamba/README.md:10]:
```suggestion
python demo/demo.py "What is the capital of France?" "What is the capital of Germany?" --model wh
```

File: models/demos/mamba/tt/model_config.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (esmalTT) [inline models/demos/mamba/tt/model_config.py:1]: We can probably delete this since it is not used and is just copied from elsewhere.

**Review by kpaigwar — COMMENTED** [review]
moved to experimental

**Review by eyonland — APPROVED** [review]

File: models/experimental/mamba/reference/tokenizer.ipynb
Comment (esmalTT) [inline models/experimental/mamba/reference/tokenizer.ipynb:1]: Should we delete this?

**Review by boris-drazic — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Pull request merged**